aptos-move/block-executor/src/scheduler.rs

```diff
@@ -143,19 +158,60 @@ impl Scheduler {
         Self {
             num_txns,
             execution_idx: AtomicUsize::new(0),
-            validation_idx: AtomicUsize::new(0),
-            decrease_cnt: AtomicUsize::new(0),
-            num_active_tasks: AtomicUsize::new(0),
+            validation_idx: AtomicU64::new(0),
+            commit_state: Mutex::new((0, 0)),
             done_marker: AtomicBool::new(false),
             txn_dependency: (0..num_txns)
                 .map(|_| CachePadded::new(Mutex::new(Vec::new())))
                 .collect(),
             txn_status: (0..num_txns)
-                .map(|_| CachePadded::new(Mutex::new(TransactionStatus::ReadyToExecute(0, None))))
+                .map(|_| {
+                    CachePadded::new((
+                        RwLock::new(ExecutionStatus::ReadyToExecute(0, None)),
+                        RwLock::new(ValidationStatus::new()),
+                    ))
+                })
                 .collect(),
         }
     }
 
+    /// If successful, returns Some(TxnIndex), the index of committed transaction.
+    /// The current implementation has one dedicated thread to try_commit.
+    pub fn try_commit(&self) -> Option<TxnIndex> {
+        let mut commit_state = self.commit_state.lock();
+        let idx = commit_state.0;
+        if idx == self.num_txns {
+            // All txns have been committed, the parallel execution can finish.
+            self.done_marker.store(true, Ordering::SeqCst);
+            return None;
+        }
+
+        if let Some(validation_status) = self.txn_status[idx].1.try_read() {
```

Comment:
there is no reason to wait on the lock, thread can go, possibly do smt useful and then come back - plus if the lock was actually held because say a new wave started, it won't be committable immediately anyway.